internal/runnerinstall/k8s.go

```diff
@@ -287,39 +288,64 @@ func (i *K8sRunnerInstaller) InstallFlags(set *flag.Set) {
 }
 
 func (i *K8sRunnerInstaller) Uninstall(ctx context.Context, opts *InstallOpts) error {
-	actionConfig, err := helminstallutil.ActionInit(opts.Log, i.Config.KubeconfigPath, i.Config.K8sContext)
+	ui := opts.UI
+	// Our checks here follow the logic of:
+	// Up until v0.8.2, we installed runners with the k8s client,
+	// and the Label was "app=waypoint-runner" and the Name "waypoint-runner-random-id"
+	// As of 0.9.0, we install runners with helm, with a Label following the
+	// pattern ("app.kubernetes.io/instance=waypoint-%s", runnerId)
+	// and the Name ("waypoint-"+strings.ToLower(runnerId))
+	//
+	// Therefore we need to ascertain A) if the runner exists on the cluster at
+	// all (it might not be if the user is auth'd to the wrong cluster), and then B)
+	// if the name/label matches the Helm pattern or the k8s client pattern so
+	// we know how to uninstall it
+
+	// A) Is runner on the cluster at all?
+	clientset, err := i.NewClient()
 	if err != nil {
+		ui.Output(err.Error(), terminal.WithErrorStyle())
 		return err
 	}
 
-	helmList := action.NewList(actionConfig)
-	releases, _ := helmList.Run()
-
-	isHelm := false
+	// Search for a runner with 0.8.x tag format, installed with k8s client
+	deploymentClient := clientset.AppsV1().Deployments(i.Config.Namespace)
+	listK8sClient, err := deploymentClient.List(ctx, metav1.ListOptions{
+		LabelSelector: fmt.Sprintf("app=%s", DefaultRunnerTagName),
+	})
+	if err != nil {
+		return fmt.Errorf("could not list deployments in namespace %q with current context: %s", i.Config.Namespace, err)
+	}
 
-	for _, release := range releases {
-		opts.Log.Debug(fmt.Sprintf("Releases %+v", release.Name))
-		if release.Name == "waypoint-"+strings.ToLower(opts.Id) {
-			isHelm = true
-			break
-		}
+	// Search for runner with 0.9+ tag format, installed with helm
+	podClient := clientset.CoreV1().Pods(i.Config.Namespace)
+	listHelmClient, err := podClient.List(ctx, metav1.ListOptions{
```

Comment:
I was wondering why this is named `listHelmClient`, but I realize now that this is looking for pods which were installed via Helm. It made me think that this was somehow used for a programmatic version of `helm list` 😄

Reply:
Ahhh yes my brain was a victim to precedence. The original variable here was `list` so then I made `listK8s` and `listHelm`. Those are confusing names! I can update them to `helmClientList` and `k8sClientList`.